File: apps/mobile/flutter/test/unit/application/use_cases/get_all_todos_use_case_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:mocktail/mocktail.dart';
import 'package:habit_todo_app/core/application/use_cases/get_all_todos_use_case.dart';
import 'package:habit_todo_app/core/domain/entities/todo_entity.dart';
import 'package:habit_todo_app/core/domain/repositories/todo_repository.dart';

class MockTodoRepository extends Mock implements TodoRepository {}

void main() {
  group('GetAllTodosUseCase', () {
    late GetAllTodosUseCase useCase;
    late MockTodoRepository mockRepository;

    setUp(() {
      mockRepository = MockTodoRepository();
      useCase = GetAllTodosUseCase(mockRepository);
    });

    test('should return all todos', () async {
      // Arrange
      final todos = [
        TodoEntity(
          id: 'todo_1',
          title: 'Todo 1',
          createdAt: DateTime(2024, 1, 1),
        ),
        TodoEntity(
          id: 'todo_2',
          title: 'Todo 2',
          createdAt: DateTime(2024, 1, 2),
        ),
      ];

      when(() => mockRepository.getAll()).thenAnswer((_) async => todos);

      // Act
      final result = await useCase.execute();

      // Assert
      expect(result, equals(todos));
      expect(result.length, 2);
      verify(() => mockRepository.getAll()).called(1);
    });

    test('should return empty list when no todos exist', () async {
      // Arrange
      when(() => mockRepository.getAll()).thenAnswer((_) async => <TodoEntity>[]);

      // Act
      final result = await useCase.execute();

      // Assert
      expect(result, isEmpty);
      verify(() => mockRepository.getAll()).called(1);
    });

    test('should delegate to repository', () async {
      // Arrange
      when(() => mockRepository.getAll()).thenAnswer((_) async => <TodoEntity>[]);

      // Act
      await useCase.execute();

      // Assert
      verify(() => mockRepository.getAll()).called(1);
    });
  });
}

File: apps/web/test/unit/application/use_cases/get_all_todos_use_case.test.ts
import { beforeEach, describe, expect, it, vi } from "vitest";
import { GetAllTodosUseCase } from "../../../../src/core/application/use_cases/get_all_todos_use_case";
import type { TodoEntity } from "../../../../src/core/domain/entities/todo_entity";
import type { TodoRepository } from "../../../../src/core/domain/repositories/todo_repository";
import { Priority } from "../../../../src/core/domain/value_objects/priority";

describe("GetAllTodosUseCase", () => {
  let mockRepository: TodoRepository;
  let useCase: GetAllTodosUseCase;

  beforeEach(() => {
    mockRepository = {
      findAll: vi.fn(),
      findById: vi.fn(),
      create: vi.fn(),
      update: vi.fn(),
      delete: vi.fn(),
    };
    useCase = new GetAllTodosUseCase(mockRepository);
  });

  it("should return all todos sorted by due date with completed at bottom", async () => {
    const todos: TodoEntity[] = [
      {
        id: "1",
        title: "Todo 1",
        priority: Priority.MEDIUM,
        dueDate: "2024-12-31T00:00:00.000Z",
        completed: false,
        createdAt: "2024-01-01T00:00:00.000Z",
        updatedAt: "2024-01-01T00:00:00.000Z",
      },
      {
        id: "2",
        title: "Todo 2",
        priority: Priority.MEDIUM,
        dueDate: "2024-12-30T00:00:00.000Z",
        completed: true,
        createdAt: "2024-01-01T00:00:00.000Z",
        updatedAt: "2024-01-01T00:00:00.000Z",
      },
      {
        id: "3",
        title: "Todo 3",
        priority: Priority.MEDIUM,
        dueDate: "2024-12-29T00:00:00.000Z",
        completed: false,
        createdAt: "2024-01-01T00:00:00.000Z",
        updatedAt: "2024-01-01T00:00:00.000Z",
      },
    ];

    vi.mocked(mockRepository.findAll).mockResolvedValue(todos);

    const result = await useCase.execute();

    expect(result.success).toBe(true);
    if (result.success) {
      expect(result.data).toHaveLength(3);
      expect(result.data[0].id).toBe("3");
      expect(result.data[1].id).toBe("1");
      expect(result.data[2].id).toBe("2");
    }
  });

  it("should return empty array when no todos", async () => {
    vi.mocked(mockRepository.findAll).mockResolvedValue([]);

    const result = await useCase.execute();

    expect(result.success).toBe(true);
    if (result.success) {
      expect(result.data).toEqual([]);
    }
  });
});




